### ISSUE TITLE
Large seeks in HTTP audio streams cause the skipped over bytes to be read (instead of using byte-range requests)

### ISSUE THREAD

**Issue by rschupp** [xbmc/xbmc#19415]
### Describe the bug

Play the following stream (3 hours long):
1. `kodi-send --action="PlayMedia(https://download-jazzfm.sharp-stream.com/jazzfm_2020307T220000-20210308T010000_somethin_else_with_jez_nelson_153909145_27.mp3?direct=true)"`
2. wait for stream to start playing, then seek ahead to the 2 hour mark:
3. `kodi-send --action="Seek(7200)"`

## Expected Behavior
After a short delay, play should resume at the new position.

## Actual Behavior

There's a delay on the order of tens of seconds until playing resumes (but the position is correct). 

Looking at the flickering lights on the ethernet hub connected to my Raspberry Pi it seems that Kodi actually reads the whole 2 hours worth of stream instead of using byte-range requests.

I checked the stream: it's a constant bit-rate encoded MP3. And the server advertises that it's capable of accepting byte-range requests. So actual "seeking" should be possible. Indeed, the same experiment   (on the same network) using VLC "seeks" with less than a second delay.

### Debuglog

The debuglog can be found here:
http://ix.io/2SOj

### Your Environment
Used Operating system:

 - [ ] Android
 - [ ] iOS
 - [ ] tvOS
 - [x] Linux
 - [ ] OSX
 - [ ] Windows
 - [ ] Windows UWP

 - Operating system version/name: Linux 5.10.17 on RPi4, Linux 5.11.6 on x86_64
 - Kodi version: 19.0 (from LibreElec master [RPi4], Debian bullseye [x86_64])

*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

**Comment by wsnipex** [comment]
> 2021-03-14 15:25:05.707 T:900     DEBUG <general>: Curl::Debug - TEXT: Mark bundle as not supporting multiuse
2021-03-14 15:25:05.707 T:900     DEBUG <general>: Curl::Debug - HEADER_IN: HTTP/1.1 206 Partial Content
2021-03-14 15:25:05.707 T:900     DEBUG <general>: Curl::Debug - HEADER_IN: Server: nginx
2021-03-14 15:25:05.707 T:900     DEBUG <general>: Curl::Debug - HEADER_IN: Date: Sun, 14 Mar 2021 14:25:05 GMT
2021-03-14 15:25:05.707 T:900     DEBUG <general>: Curl::Debug - HEADER_IN: Content-Type: audio/mpeg
2021-03-14 15:25:05.707 T:900     DEBUG <general>: Curl::Debug - HEADER_IN: Content-Length: 131072
2021-03-14 15:25:05.707 T:900     DEBUG <general>: Curl::Debug - HEADER_IN: Last-Modified: Wed, 10 Mar 2021 11:54:22 GMT
2021-03-14 15:25:05.707 T:900     DEBUG <general>: Curl::Debug - HEADER_IN: Connection: keep-alive
2021-03-14 15:25:05.707 T:900     DEBUG <general>: Curl::Debug - HEADER_IN: ETag: "6048b36e-1a6ad629"
2021-03-14 15:25:05.707 T:900     DEBUG <general>: Curl::Debug - HEADER_IN: Access-Control-Allow-Origin: *
2021-03-14 15:25:05.707 T:900     DEBUG <general>: Curl::Debug - HEADER_IN: Content-Range: bytes 443078185-443209256/443209257

Kodi clearly requests the content range. I guess what happens is that the returned data is way too small:
443209257 - 443208104 = 1153 bytes

**Comment by rschupp** [comment]
@wsnipex I think you actually wanted to refer to the line:
```
2021-03-14 15:25:05.708 T:737     DEBUG <general>: Seek - <https://download-jazzfm.sharp-stream.com/jazzfm_20210307T220000-20210308T010000_somethin_else_with_jez_nelson_153909145_27.mp3?direct=true> waiting for position 443208104
```
But that line is **not** caused by "seek to 7200 secs" - it's always there, just by playing the stream from the start (dunno why Kodi does that - perhaps checking for trailing ID3v1 data?). 

And for "seek to 7200 secs" that byte offset is way off:
- file has constant bitrate = 320 kb/s, samplerate = 44100 Hz
- resulting in frame length = floor( 144 * 320000 / 44100 +1 ) = 1045
- length of ID3v2 header = 8151
- hence estimated byte offset for 7200 sec =  8151 + 1045 * 7200 / 0.026 =  2546584615

**Comment by Tolriq** [comment]
@wsnipex as more users update to Kodi 19 I start to get quite a few report of same kind of issues. They try to stream from my app to Kodi and Kodi waits an insane long time before playing or even can't for large files. (I serve the file via a proper web server).

From Yatse side logs, Kodi does indeed read the whole file.

The relevant part of Kodi logs:

```
2021-03-23 09:13:39.982 T:30352    INFO <general>: Creating InputStream
2021-03-23 09:13:39.982 T:30352   DEBUG <general>: CurlFile::Open(0x1defc9eb30) http://192.168.1.170:39903/F76DA67E1734E9F55BA00DFBA014D3C2/%2Fstorage%2Femulated%2F0%2FDCIM%2FCamera%2FPXL_20210209_120357735.mp4#FromYatse
2021-03-23 09:13:39.984 T:31784   DEBUG <general>: Loading settings for http://192.168.1.170:39903/F76DA67E1734E9F55BA00DFBA014D3C2/%2Fstorage%2Femulated%2F0%2FDCIM%2FCamera%2FPXL_20210209_120357735.mp4#FromYatse
2021-03-23 09:13:39.985 T:41832   DEBUG <general>: CVideoGUIInfo::InitCurrentItem(http://192.168.1.170:39903/F76DA67E1734E9F55BA00DFBA014D3C2/%2Fstorage%2Femulated%2F0%2FDCIM%2FCamera%2FPXL_20210209_120357735.mp4#FromYatse)
2021-03-23 09:13:39.990 T:41832   DEBUG <general>: CPlayerGUIInfo::InitCurrentItem(http://192.168.1.170:39903/F76DA67E1734E9F55BA00DFBA014D3C2/%2Fstorage%2Femulated%2F0%2FDCIM%2FCamera%2FPXL_20210209_120357735.mp4#FromYatse)
2021-03-23 09:13:39.992 T:40584   DEBUG <general>: CVideoDatabase::GetMovieId (http://192.168.1.170:39903/F76DA67E1734E9F55BA00DFBA014D3C2/%2Fstorage%2Femulated%2F0%2FDCIM%2FCamera%2FPXL_20210209_120357735.mp4#FromYatse), query = select idMovie from movie where idFile=15831
2021-03-23 09:13:39.992 T:40584   DEBUG <general>: CVideoDatabase::GetEpisodeId (http://192.168.1.170:39903/F76DA67E1734E9F55BA00DFBA014D3C2/%2Fstorage%2Femulated%2F0%2FDCIM%2FCamera%2FPXL_20210209_120357735.mp4#FromYatse), query = select idEpisode from episode where idFile=15831
2021-03-23 09:13:39.992 T:40584   DEBUG <general>: CVideoDatabase::GetMusicVideoId (http://192.168.1.170:39903/F76DA67E1734E9F55BA00DFBA014D3C2/%2Fstorage%2Femulated%2F0%2FDCIM%2FCamera%2FPXL_20210209_120357735.mp4#FromYatse), query = select idMVideo from musicvideo where idFile=15831
2021-03-23 09:13:39.993 T:40584   DEBUG <UPNP::BuildObject>: Building didl for object 'http://192.168.1.170:39903/F76DA67E1734E9F55BA00DFBA014D3C2/%2Fstorage%2Femulated%2F0%2FDCIM%2FCamera%2FPXL_20210209_120357735.mp4#FromYatse'
2021-03-23 09:13:40.012 T:30352   DEBUG <general>: CFileCache::Open - opening <F76DA67E1734E9F55BA00DFBA014D3C2/%2Fstorage%2Femulated%2F0%2FDCIM%2FCamera%2FPXL_20210209_120357735.mp4> using cache
2021-03-23 09:13:40.012 T:30352   DEBUG <general>: CurlFile::Open(0x14df615f530) http://192.168.1.170:39903/F76DA67E1734E9F55BA00DFBA014D3C2/%2Fstorage%2Femulated%2F0%2FDCIM%2FCamera%2FPXL_20210209_120357735.mp4#FromYatse
2021-03-23 09:13:40.013 T:3268    DEBUG <general>: Twitch: Player: |onPlayBackStarted| isTwitch |False| SeekTime ||
2021-03-23 09:13:40.034 T:30352   DEBUG <general>: CFileCache::Open - Source chunk size is 0, setting cache chunk size to 131072
2021-03-23 09:13:40.034 T:30352   DEBUG <general>: CFileCache::Open - Using double memory cache each sized 10485760 bytes
2021-03-23 09:13:40.035 T:22340   DEBUG <general>: Thread FileCache start, auto delete: false
2021-03-23 09:13:40.035 T:30352   DEBUG <general>: CUtil::ScanForExternalSubtitles: Searching for subtitles...
2021-03-23 09:13:40.035 T:30352   DEBUG <general>: CurlFile::Open(0x1defc9c390) http://192.168.1.170:39903/F76DA67E1734E9F55BA00DFBA014D3C2/
2021-03-23 09:13:40.035 T:30352   DEBUG <general>: XCURL::DllLibCurlGlobal::easy_acquire - Created session to http://192.168.1.170

2021-03-23 09:14:24.048 T:30352   DEBUG <general>: CUtil::ScanForExternalSubtitles: END (total time: 44014 ms)
2021-03-23 09:14:24.048 T:30352    INFO <general>: Creating Demuxer
2021-03-23 09:14:24.048 T:30352   DEBUG <general>: CDVDDemuxFFmpeg::Open - probing detected format [mov,mp4,m4a,3gp,3g2,mj2]
2021-03-23 09:14:24.050 T:22340   DEBUG <general>: CurlFile::CReadState::Connect - Resume from position 639832165
2021-03-23 09:14:24.125 T:22340   DEBUG <general>: CFileCache::Process - Cache completely reset for seek to position 639832165
2021-03-23 09:14:24.125 T:30352   DEBUG <general>: XFILE::CFileCache::Seek - waiting for position 639840614.
2021-03-23 09:14:24.134 T:22340   DEBUG <general>: CFileCache::Process - Source read didn't return any data! Hit eof(?)
2021-03-23 09:14:24.144 T:30352   DEBUG <general>: CDVDDemuxFFmpeg::Open - avformat_find_stream_info starting
2021-03-23 09:14:24.360 T:30352   DEBUG <general>: CDVDDemuxFFmpeg::Open - av_find_stream_info finished
```

----> CUtil::ScanForExternalSubtitles: END (total time: 44014 ms) 44 seconds here. 
----> CurlFile::CReadState::Connect - Resume from position 639832165 (End of file)


The request that Yatse server gets:
```
2021-03-23 09:13:40.196 10486-10530/org.leetzone.android.yatsewidgetfree.debug V/StreamingServer: StreamingServer.serve@239: Request: [GET] /F76DA67E1734E9F55BA00DFBA014D3C2//storage/emulated/0/DCIM/Camera/PXL_20210209_120357735.mp4 {host=192.168.1.170:39903, range=bytes=0-, accept-encoding=deflate, gzip, br, user-agent=Kodi/19.0-RC1 (Windows NT 10.0.19042.867; Win64; x64) App_Bitness/64 Version/19.0-RC1-(18.9.910)-Git:20210115-90a1e12804, accept=*/*, accept-charset=UTF-8,*;q=0.8}
2021-03-23 09:13:40.227 10486-10540/org.leetzone.android.yatsewidgetfree.debug V/StreamingServer: StreamingServer.serve@239: Request: [GET] /F76DA67E1734E9F55BA00DFBA014D3C2//storage/emulated/0/DCIM/Camera/PXL_20210209_120357735.mp4 {host=192.168.1.170:39903, range=bytes=0-, accept-encoding=deflate, gzip, br, user-agent=Kodi/19.0-RC1 (Windows NT 10.0.19042.867; Win64; x64) App_Bitness/64 Version/19.0-RC1-(18.9.910)-Git:20210115-90a1e12804, accept=*/*, accept-charset=UTF-8,*;q=0.8}
2021-03-23 09:13:40.254 10486-10533/org.leetzone.android.yatsewidgetfree.debug V/StreamingServer: StreamingServer.serve@239: Request: [GET] /F76DA67E1734E9F55BA00DFBA014D3C2/ {host=192.168.1.170:39903, range=bytes=0-, accept-encoding=deflate, gzip, br, user-agent=Kodi/19.0-RC1 (Windows NT 10.0.19042.867; Win64; x64) App_Bitness/64 Version/19.0-RC1-(18.9.910)-Git:20210115-90a1e12804, accept=*/*, accept-charset=UTF-8,*;q=0.8}
2021-03-23 09:14:24.283 10486-10540/org.leetzone.android.yatsewidgetfree.debug V/StreamingServer: StreamingServer.serve@239: Request: [GET] /F76DA67E1734E9F55BA00DFBA014D3C2//storage/emulated/0/DCIM/Camera/PXL_20210209_120357735.mp4 {host=192.168.1.170:39903, range=bytes=639832165-, accept-encoding=deflate, gzip, br, user-agent=Kodi/19.0-RC1 (Windows NT 10.0.19042.867; Win64; x64) App_Bitness/64 Version/19.0-RC1-(18.9.910)-Git:20210115-90a1e12804, accept=*/*, accept-charset=UTF-8,*;q=0.8}

```

Basically Kodi reads the whole file to it's cache then plays from it's cache.

**Comment by Tolriq** [comment]
ping @arnova as seems he's the last to have PR on cache stuff.

**Comment by wsnipex** [comment]
> ----> CUtil::ScanForExternalSubtitles: END (total time: 44014 ms) 44 seconds here.

 :roll_eyes:

**Comment by arnova** [comment]
> ping @arnova as seems he's the last to have PR on cache stuff.

I don't think it's a cache issue. The actual problem seems to be a non-seekable stream in combination with "something" that wants to read at eof (tag?). I recall some stuff was changed in that area, I'll see whether I can find any commits that might have triggered this.

**Comment by arnova** [comment]
I think this is the offending one: https://github.com/xbmc/xbmc/pull/18218

**Comment by Tolriq** [comment]
So after @wsnipex eloquent comment I reverted https://github.com/xbmc/xbmc/pull/17491 (so not more sub checks) and issue is fixed.
What's still strange is that that function correctly request the folder on Yatse that answer instantly and I do not see anything in that function that would make the file be read in it's entirety.
Furthermore after Kodi no more makes any request to the webserver so it's using it's cache somehow. (Unless I manually seek that works nicely)

So as the file is seekable I'm not sure my issue is really similar to OP.

**Comment by arnova** [comment]
> So after @wsnipex eloquent comment I reverted #17491 (so not more sub checks) and issue is fixed.
> What's still strange is that that function correctly request the folder on Yatse that answer instantly and I do not see anything in that function that would make the file be read in it's entirety.
> Furthermore after Kodi no more makes any request to the webserver so it's using it's cache somehow. (Unless I manually seek that works nicely)
> 
> So as the file is seekable I'm not sure my issue is really similar to OP.

Yes, I think it's a different issue. The OPs concerns an internet audio stream, yours is a local video stream, right?

**Comment by rschupp** [comment]
> I think this is the offending one: #18218

OP here, I just retried with #18218 reverted, but to no avail.

**Comment by arnova** [comment]
> > I think this is the offending one: #18218
> 
> OP here, I just retried with #18218 reverted, but to no avail.

Hmmm, yes that would have been way too easy ;-) Could you figure out when this problem was (roughly) introduced (Matrix alphas/beta/rc) ?

**Comment by PenOnFace** [comment]
Just to add, I am getting this with video files streamed from a HTTP server, but only on Android (Chromecast with Google TV). The Windows version seems fine and sends correct byte ranges to load the index at the end of the file first, to allow instant playback - and skipping around the video using ranges - without having to download the whole file.

The Android version keeps sending "Range: bytes=0-" instead. This has been in place from Matrix v19.0 and was working perfectly on Android when I was using v18.9 (and still is when I sideload the old APK). Is there a different version of libcurl used on Android compared to Windows?

**Comment by PenOnFace** [comment]
Just a small update on this... The behaviour I was seeing was due to Kodi trying to search for subtitles by removing the last part of the path (the filename) from the URL to try and return a directory listing.

The HTTP server I use was set to return a large (several GB) default file for that directory, so it looks like Kodi was trying to download that to parse it, before loading in the rest of the file.

If I force the server to return a 404 not found when trying to load a directory listing, Kodi continues on with the load correctly after the subtitle check and does a final connect to load the end of the file using a range request. Playback then commences immediately.

The log says this just before the connection that makes it freeze:
DEBUG <general>: ScanForExternalSubtitles: Searching for subtitles...

I don't know if this is expected behaviour to try and scan a folder when it's connecting to a HTTP server?
